temporalio/worker/_tuning.py

```diff
@@ -0,0 +1,185 @@
+            max_slots,
+            int(ramp_throttle / timedelta(milliseconds=1)),
+            temporalio.bridge.worker.ResourceBasedTunerOptions(
+                slot_supplier.tuner_options.target_memory_usage,
+                slot_supplier.tuner_options.target_cpu_usage,
+            ),
+        )
+    else:
+        raise TypeError(f"Unknown slot supplier type: {slot_supplier}")
+
+
+class WorkerTuner(ABC):
+    """WorkerTuners allow for the dynamic customization of some aspects of worker configuration"""
+
+    @abstractmethod
+    def _get_workflow_task_slot_supplier(self) -> SlotSupplier:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _get_activity_task_slot_supplier(self) -> SlotSupplier:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _get_local_activity_task_slot_supplier(self) -> SlotSupplier:
+        raise NotImplementedError
+
+
+class ResourceBasedTuner(WorkerTuner):
+    """This tuner attempts to maintain certain levels of resource usage when under load.
+
+    .. warning::
+        The resource based tuner is currently experimental.
+    """
+
+    def __init__(self, options: ResourceBasedTunerOptions):
+        """Create a new ResourceBasedTuner
+
+        Args:
+            options: Specify the target resource usage levels for the tuner
+        """
+        self.options = options
+        self.workflow_task_options: Optional[ResourceBasedSlotOptions] = None
+        self.activity_task_options: Optional[ResourceBasedSlotOptions] = None
+        self.local_activity_task_options: Optional[ResourceBasedSlotOptions] = None
+
+    def set_workflow_task_options(self, options: ResourceBasedSlotOptions):
+        """Set options for the workflow task slot supplier"""
+        self.workflow_task_options = options
+
+    def set_activity_task_options(self, options: ResourceBasedSlotOptions):
+        """Set options for the activity task slot supplier"""
+        self.activity_task_options = options
+
+    def set_local_activity_task_options(self, options: ResourceBasedSlotOptions):
+        """Set options for the local activity task slot supplier"""
+        self.local_activity_task_options = options
+
+    def _get_workflow_task_slot_supplier(self) -> SlotSupplier:
+        return ResourceBasedSlotSupplier(
+            self.workflow_task_options or ResourceBasedSlotOptions(None, None, None),
+            self.options,
+        )
+
+    def _get_activity_task_slot_supplier(self) -> SlotSupplier:
+        return ResourceBasedSlotSupplier(
+            self.activity_task_options or ResourceBasedSlotOptions(None, None, None),
+            self.options,
+        )
+
+    def _get_local_activity_task_slot_supplier(self) -> SlotSupplier:
+        return ResourceBasedSlotSupplier(
+            self.local_activity_task_options
+            or ResourceBasedSlotOptions(None, None, None),
+            self.options,
+        )
+
+
+@dataclass(frozen=True)
+class CompositeTuner(WorkerTuner):
```

Comment:
I am now thinking that this class doesn't need to be exposed either but instead can also be a static method on worker tuner that accepts these three kwargs and returns a worker tuner. There is no value from a user POV to expose a new type vs just a method to make a worker tuner from these three values.